c++/src/capnp/serialize-async.c++

```diff
@@ -541,4 +545,181 @@ kj::Promise<MessageReaderAndFds> MessageStream::readMessage(
   });
 }
 
+// TODO(tomorrow):
+//
+// Handling retainLastMessage() is awkward with it on MessageStream instead of a returned message,
+// since now messageStream needs to be aware of whether or not the last message it returned
+// was inside of the buffer, or used a temporary overflow buffer. I think this is justification
+// for making BufferedMessageStream an entirely separate type.
+//
+// How should we use scratchSpace? there's really two opportunities
+// * on read, to use for overflow space
+//   This seems like it's of dubious usefulness
+// * on retain -- this would allow allocating space inside of the rpc call context to hold small messages
+//   This would be useful, and hits the more common case
+//
+
+// Returned kj::Own<MessageReader>s need to keep the buffer around
+// "retaining" a message
+
+class BufferedAsyncMessageStream final : public BufferedMessageStream {
```

Comment:
I'm guessing this lives in serialize-async.c++ because buffering.c++ is in kj, right? Just checking because it feels natural to have it next to the buffering implementation it uses.